core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -509,6 +511,13 @@ private[spark] class ExecutorAllocationManager(
   private def onExecutorBusy(executorId: String): Unit = synchronized {
     logDebug(s"Clearing idle timer for $executorId because it is now running a task")
     removeTimes.remove(executorId)
+
+    // Executor is added to remove by misjudgment due to async listener making it as idle).
+    // see SPARK-9552
+    if (executorsPendingToRemove.contains(executorId)) {
```

Comment:
So, backtracking a step because this code can be confusing.

`executorsPendingToRemove` is only updated by `removeExecutor` in this class. `removeExecutor` only adds an executor to that list if calls `ExecutorAllocationClient.killExecutor` and the result is `true`, meaning a request was made to kill the executor.

So if you do what I suggested above (change the return value of `ExecutorAllocationClient.killExecutor` so that it returns `false` when the executor is busy, and thus a request to kill it was not sent), the executor would not be added to this list in the first place, and you wouldn't need to fix it.

Does that make sense? Why can't that be done?